language/move-vm/runtime/src/data_cache.rs

```diff
@@ -197,7 +199,20 @@ impl<'r, 'l, C: RemoteCache> DataStore for TransactionDataCache<'r, 'l, C> {
 
                     GlobalValue::cached(val)?
                 }
-                None => GlobalValue::none(),
+                Ok(None) => GlobalValue::none(),
+                Err(err) => {
+                    println!("HHIIIIIT");
+                    let msg = format!("Unexpected storage error: {:?}", err);
+                    // REVIEW: better way to get info out of a PartialVMError?
+                    let (_old_status, _old_sub_status, _old_message, _location, indices, offsets) =
+                        err.finish(Location::Undefined).all_data();
+                    return Err(
+                        PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
```

Comment:
Why is this an invariant violation? I don't doubt it, just am confused

Reply:
Here's my reasoning: when the storage backend returns you an error, it means something is really really wrong in the sense that you have no idea what has happened at all, and you want to fail fast and discard the transaction. This is not a user error and you should never charge the user for it.

IMO what makes this less obvious and error prone is that `trait RemoteCache` requires the storage to use `VMError` and `PartialVMError`, the same stuff the VM uses internally itself, instead of a dedicated `VMStorageError` or maybe an associated error type. This enables a shortcut where the VM can just propagates whatever errors the storage returns without handling them, which is very wrong.

Reply:
This makes sense.
Maybe give it it's own invariant violation code? Or a bucket for storage invariant violations? Fine either way though

Reply:
> a bucket for storage invariant violations
I think this is definitely something we should keep thinking about. At the moment I don't think we fully understand what kind of interfaces we want for storage integration, and if the there's anything special about storage to justify a separate bucket of invariant violations. Also I feel like it's a little bit too late to introduce changes like this.

> give it it's own invariant violation code?
We could, but I guess it's not high priority at the moment -- the Libra VM only cares about the bucket so as long as the error returned is in the correct bucket, we're good. Error codes in the invariant violation bucket are a little bit inconsistent and unorganized in general, so we'll definitely revisit them later.